content/blog/2019-11-02-custom-alloc/index.md

```diff
@@ -0,0 +1,206 @@
+
+
+#### Saving Space with Custom Allocation
+
+The example above actually uses more memory than a heap. That is more common, but they can also be used to reduce the memory footprint, primarily by preventing fragmentation. For instance, if memory fragmentation is a huge issue, but you know rougly what fraction of your memory is used by objects of different sizes, you can partition it into pieces, and deal with each size class separately, which offers much better guarantees about the worst and average case fragemtation. This is commonly done in practice; [Mesh][], for instance, requires that all pages in question are of the same size class. [The paper we analyze][paper] refers to allocators liek these as "per-class custom allocators".
+
+For all of these reasons, in 2002, it was common practice and widely condisered a good idea to write a custom memory allocation for your program to improve performance. 
+
+
+## The Drawbacks to Custom Allocation
+
+Of course, even without any experiments, it is easy to see how doing this could be a mistake: the effectiveness of any of these relies on certain assumptions about the programs, and so any modifications need to be done while keeping the custom allocator in mind. Moreover, non-standard allocation schemes make it much more difficult to use existing tools to analyze your code. In more detail:
+
+* Accidentally calling the standard `free` on a custom-allocated object could corrupt the heap and lead to errors
+* Custom memory allocation makes it impossible to use leak detection tools
+* It also prevents you from using a different custom allocator to do something else that's more valuable in the future
+* It keeps you from using garbage collection
+
+
+
+# The Paper: Reconsidering Custom Memory Management
+
+The paper at hand, [Reconsidering Custom Memory Allocation][paper], is packaged and sold as the thesis that custom memory allocators in practice do not out perform general ones. 
+
+One of the key arguments of this paper is that standard baselines are not fair. Evidently the usual argument _in favor of_ cusom allocation in 2002 was a comparison against the win32 memory allocator, which was much slower than custom allocators. The first part of this paper is an evaluation against the Lea allocator (Doug Lea's `dlmalloc`, now often the default allocator), which greatly reduces and in some cases eliminates the margin of victory for custom allocators not exploting regions. 
+
+
+### The Taxonomy of Memory Allocators.
+This paper uses the following three bins to classify allocators: 
+
+**[1] Per-class allocators**. You build a separate heap for every object size class, and optimize each one separately for objects of this size. This is fast, simple, and interacts with C well --- but could be space inefficient if you don't know how much memory of each class size you will use.
```

Comment:
No need for spaces around em dashes. http://www.read.seas.harvard.edu/%7Ekohler/latex.html#dash